Create unit tests for `catalog/db.py` and add CI for tests

nahumsa opened the pull request:
Related to #2 

This will add CI and tests for `catalog/db.py`.

nahumsa commented on tests/catalog/test_db.py:
```diff
@@ -0,0 +1,102 @@
+        catalog.touch("/databases/table")
+        catalog.rm("/databases/table")
+
+        with catalog.Session() as session:
+            entry = session.query(CatalogEntry).filter_by(parent="/databases", name="table").first()
+            assert entry.is_deleted
+
+        assert catalog.ls("/databases/table") == None
```
I am not sure which way to test this is better. I think that testing using `catalog.ls` is a good way because it becomes somewhat agnostic and can be used for other instances of `AbstractCatalog`. But maybe having both tests is a good thing too.

criccomini replied:
I agree `.ls()` is the best way. As commented above, I suggest you use *all* of the catalog's methods (ls, rm, read, touch, etc). This way, the test becomes completely portable and you can matrix in all catalogs. They should all behave the same.

nahumsa commented on tests/catalog/test_db.py:
```diff
@@ -0,0 +1,102 @@
+                "table": "some_table"
+            }
+        }
+        path = "/foo/bar/baz"
```
Should we use a more "real world" like path, or is this fine?

criccomini replied:
Just to make things less confusing, I think using a real path would be better. In this case:

    /databases/postgresql/instances/localhost/schemas/some_db/tables/some_table

criccomini commented on tests/catalog/test_db.py:
```diff
@@ -0,0 +1,102 @@
+        entry = CatalogEntry(deleted_at=None)
+        assert entry.is_deleted() == False
+
+class TestDatabaseCatalog:
```
Nit: I've been doing 2 blank lines between classes to play nicely with linters.

nahumsa replied:
In other projects I usually use black and isort for code style.
Should we open an issue to discuss linters and code formatting or is it to soon?

criccomini replied:
Yea, I think it's worth opening an issue for it. Sooner the better.

criccomini commented on tests/catalog/test_db.py:
```diff
@@ -0,0 +1,102 @@
+
+    def test_catalog_touch(self, catalog):
+        # Test touching a path that doesn't exist
+        catalog.touch("/databases/table")
```
Nit: To make things slightly less confusing, can you use a realistic path, eg:

    /databases/postgresql/instances/localhost

criccomini commented on tests/catalog/test_db.py:
```diff
@@ -0,0 +1,102 @@
+    def test_catalog_touch(self, catalog):
+        # Test touching a path that doesn't exist
+        catalog.touch("/databases/table")
+        with catalog.Session() as session:
```
Rather than sneaking into the `.Session()` instance variable, can you use `ls()` here to see if the child was created? It's a bit safer not to depend on internal implementation details for the test.

criccomini commented on tests/catalog/test_db.py:
```diff
@@ -0,0 +1,102 @@
+            assert result.metadata_ == {}
+
+        # Test touching a path that has been deleted
+        with catalog.Session() as session:
```
Same deal here: suggest using `rm()`

criccomini commented on tests/catalog/test_db.py:
```diff
@@ -0,0 +1,102 @@
+            entry.deleted_at = datetime.now()
+
+        catalog.touch("/databases/table")
+        with catalog.Session() as session:
```
And `children()` or `read()` here

criccomini commented on tests/catalog/test_db.py:
```diff
@@ -0,0 +1,102 @@
+
+    def test_write(self, catalog):
+        metadata = {
+            "db.location": {
```
Can you use the actual Pydantic `Location` model here?

nahumsa replied:
Done

criccomini commented on tests/catalog/test_db.py:
```diff
@@ -0,0 +1,102 @@
+
+    def test_rm(self, catalog):
+        catalog.touch("/databases/table")
+        catalog.rm("/databases/table")
```
For paranoia, should verify that path exists after touch, before rm.

criccomini commented on tests/catalog/test_db.py:
```diff
@@ -0,0 +1,102 @@
+
+    def test_ls_one_entry(self, catalog):
+        catalog.write('/databases/schema/table_one', {})
+        assert set(catalog.ls('/databases/schema')) == set(['table_one'])
```
To be *extra* paranoid here, I suggest not using set. This way, you will detect if ls() returns duplicates for a child (e.g. ['table_one', 'table_one']).

criccomini commented on tests/catalog/test_db.py:
```diff
@@ -0,0 +1,102 @@
+        assert set(catalog.ls('/databases/schema')) == set(['table_one','table_two','table_three'])
+
+    def test_search(self, catalog):
+        metadata = {
```
Same Pydantic model comment as above.

criccomini commented:
This is looking like a great start. Really appreciate you spending time on this! =) I left some comments above.

nahumsa commented:
Thanks for the quick reply! I'll address all the comments. :)

criccomini commented:
@nahumsa FYI, you might want to rebase. I just pushed a fix:

https://github.com/recap-cloud/recap/pull/156

After discovering a bug in `DatabaseCatalog.rm()` that would probably cause some of your tests to fail.

nahumsa commented:
@criccomini I think that the PR is ready for another review. I addressed all the comments.
I just need to rebase, and I think it's good to be merged! :)

criccomini commented:
Awesome, will take a look!

criccomini commented:
@nahumsa could you resolve the merge conflict? I think pdm.lock needs an update.

criccomini commented on tests/catalog/test_db.py:
```diff
@@ -0,0 +1,120 @@
+    def catalog(self, engine):
+        return DatabaseCatalog(engine)
+
+    def test_catalog_init(self, catalog):
```
Nit: remove. This is testing an implementation detail of the catalog.

criccomini commented on tests/catalog/test_db.py:
```diff
@@ -0,0 +1,120 @@
+        parent_path = Path("/databases/postgresql/instances/localhost/schemas/some_db/tables/")
+        child_path = Path("some_table")
+
+        catalog.touch(parent_path / child_path)
```
Woah. Mind blown. Didn't know you could do this with Path!

criccomini commented on tests/catalog/test_db.py:
```diff
@@ -0,0 +1,120 @@
+        catalog.touch(parent_path / child_path)
+        assert catalog.ls(parent_path) == [str(child_path)]
+
+    @pytest.mark.skip(reason="Waiting to rebase main")
```
I think this needs to be removed

criccomini commented on tests/catalog/test_db.py:
```diff
@@ -0,0 +1,120 @@
+        parent_path = Path("/databases/postgresql/instances/localhost/schemas/some_db/tables")
+        child_path = Path("table_one")
+        catalog.write(parent_path / child_path, {})
+        assert sorted(catalog.ls(parent_path)) == sorted([str(child_path)])
```
Curious why `sorted` here if one entry?

nahumsa replied:
It was just to have a consistent pattern when using `catalog.ls`. If you think that this is confusing, I can remove it.
It works without the `sorted`.

criccomini commented on tests/catalog/test_db.py:
```diff
@@ -0,0 +1,120 @@
+        catalog.write(parent_path / child_path, metadata.dict(), patch=False)
+        search_result = catalog.search(
+            "json_extract(metadata, '$.\"database\"') = 'db'"
+            )
```
Nit, `)` is one indent too far I think.

criccomini reviewed (commented):
Ok, did another pass. Really close, just some nits and questions (and the rebase). Thanks again for this. Looks really great.

nahumsa commented:
Thanks @criccomini! I'll address all the comments, rebase and resolve the `pdm.lock` conflict.

nahumsa commented:
@criccomini I have just rebased and resolved  the conflicts on the `pdm.lock` file. I also added a new test `test_write_metadata_after_rm`.

criccomini commented:
🎊 thanks! Merged =)
